File: openHAB/NSData+HexString.m
//
//  NSData+NSData_HexString.m
//  openHAB
//
//  Created by Victor Belov on 05/04/14.
//  Copyright (c) 2014 Victor Belov. All rights reserved.
//

#import "NSData+HexString.h"

@implementation NSData (HexString)

- (NSString *)hexString {
    /* Returns hexadecimal string of NSData. Empty string if data is empty.   */
    
    const unsigned char *dataBuffer = (const unsigned char *)[self bytes];
    
    if (!dataBuffer)
        return [NSString string];
    
    NSUInteger          dataLength  = [self length];
    NSMutableString     *hexString  = [NSMutableString stringWithCapacity:(dataLength * 2)];
    
    for (int i = 0; i < dataLength; ++i)
        [hexString appendString:[NSString stringWithFormat:@"%02lx", (unsigned long)dataBuffer[i]]];
    
    return [NSString stringWithString:hexString];
}

@end

File: openHAB/NSData+HexString.swift
//  Converted to Swift 4 by Swiftify v4.2.20229 - https://objectivec2swift.com/
//
//  NSData+NSData_HexString.h
//  openHAB
//
//  Created by Victor Belov on 05/04/14.
//  Copyright (c) 2014 Victor Belov. All rights reserved.
//

import Foundation

extension Data {
    /// A hexadecimal string representation of the bytes.
    func hexString() -> String {
        let hexDigits = Array("0123456789abcdef".utf16)
        var hexChars = [UTF16.CodeUnit]()
        hexChars.reserveCapacity(count * 2)

        for byte in self {
            let (index1, index2) = Int(byte).quotientAndRemainder(dividingBy: 16)
            hexChars.append(hexDigits[index1])
            hexChars.append(hexDigits[index2])
        }

        return String(utf16CodeUnits: hexChars, count: hexChars.count)
    }
}
